distributed/active_memory_manager.py

```diff
@@ -235,32 +264,87 @@ def _find_dropper(
         The worker with the highest memory usage (downstream of pending replications and
         drops), or None if no eligible candidates are available.
         """
+        orig_candidates = candidates
+
+        def log_reject(msg: str) -> None:
+            task_logger.debug("(drop, %s, %s) rejected: %s", ts, orig_candidates, msg)
+
         if len(ts.who_has) - len(pending_drop) < 2:
-            logger.debug(
-                "(drop, %s, %s) rejected: less than 2 replicas exist", ts, candidates
-            )
+            log_reject("less than 2 replicas exist")
             return None
+
         if candidates is None:
             candidates = ts.who_has.copy()
         else:
-            candidates &= ts.who_has
+            # Don't modify orig_candidates
+            candidates = candidates & ts.who_has
+            if not candidates:
+                log_reject("no candidates suggested by the policy own a replica")
+                return None
+
         candidates -= pending_drop
-        candidates -= {waiter_ts.processing_on for waiter_ts in ts.waiters}
         if not candidates:
-            logger.debug("(drop, %s, %s) rejected: no valid candidates", ts, candidates)
+            log_reject("already pending drop on all candidates")
+            return None
+
+        # The `candidates &` bit could seem redundant with `candidates -=` immediately
+        # below on first look, but beware of the second use of processing_on later on!
```

Comment:
```suggestion
        # below on first look, but beware of the second use of this variable later on!
```
didn't write out the whole name because it's very long